build.sh

```diff
@@ -26,26 +26,31 @@ shift # past argument or value
 done
 
 # Flatten resulting image.
+# shellcheck disable=SC2016
 function flatten() {
   local image="$1"
   echo "Flattening image '$image'..."
-
-  # Since `--squash` is still experimental, we have to flatten the image
-  # by exporting and importing a container based on the source image. By
-  # doing this though, we lose a lot of the Dockerfile variables that are
-  # required for running the image (ENV, EXPOSE, WORKDIR, etc) so we
-  # manually rebuild them.
-  # See here for more details: https://github.com/moby/moby/issues/8334
   local container
   container=$(docker create "$image")
+  local envs
+  envs=$(docker inspect -f '{{range $index, $value := .Config.Env}}{{$value}} {{end}}' "$container")
+  local workDir
+  workDir=$(docker inspect -f '{{ .Config.WorkingDir }}' "$container")
+  local user
+  user=$(docker inspect -f '{{ .Config.User }}' "$container")
+  local entrypoint
+  entrypoint=$(docker inspect -f '[{{range $index, $value := .Config.Entrypoint }}{{if $index}},{{end}}"{{$value}}"{{end}}]' "$container")
+  local cmd
+  cmd=$(docker inspect -f '[{{range $index, $value := .Config.Cmd }}{{if $index}},{{end}}"{{$value}}"{{end}}]' "$container")
+  local ports
+  IFS=":" read -r -a ports <<< "$(docker inspect -f '{{range $port, $empty := .Config.ExposedPorts}}--change:EXPOSE {{$port}}:{{end}}' "$container")"
```

Comment:
This is clever and a nice improvement to prevent duplicating configuration. Mind adding a comment at the top explaining what this is doing so it's easier to understand at-a-glance?

Reply:
Added some note.